src/main/java/at/hannibal2/skyhanni/features/inventory/FannCost.kt

```diff
@@ -0,0 +1,258 @@
+     * REGEX-TEST: User Input
+     */
+    private val userInputPattern by patternGroup.pattern(
+        "slot24.name.input",
+        "User Input",
+    )
+
+    /**
+     * REGEX-TEST: Type: Free
+     * REGEX-TEST: Type: Light
+     * REGEX-TEST: Type: Moderate
+     * REGEX-TEST: Type: Expert
+     * REGEX-TEST: Type: Ultra
+     * REGEX-TEST: Type: Turbo!
+     */
+    private val trainingTypePattern by patternGroup.pattern(
+        "training.type",
+        "Type: (Free|Light|Moderate|Expert|Ultra|Turbo!)",
+    )
+
+    @SubscribeEvent
+    fun onFannAnvilTooltip(event: LorenzToolTipEvent) {
+        if (!trainingSlotInventoryPattern.matches(InventoryUtils.openInventoryName())) return
+        if (!anvilPattern.matches(event.itemStack.displayName)) return
+        val tooltip = event.toolTip
+
+        val trainingType = tooltip.getTrainingType() ?: return
+        if (trainingType == TrainingType.FREE) return
+        if (!showCoins || !showBits) return
+
+        when (trainingMode) {
+            TrainingMode.DAY_COUNT -> {
+                val totalExp = tooltip.getExpEarned() ?: return
+                val coinPerExp = tooltip.getCoins() / totalExp
+                val xpPerBit = totalExp / tooltip.getBits()
+
+                tooltip.insertLineAfter(coinsPattern, "§6 ➜ Coins/XP: ${coinPerExp.roundTo(2)}")
+                tooltip.insertLineAfter(bitsPattern, "§b ➜ XP/Bit: ${xpPerBit.roundTo(2)}")
+
+            }
+
+            TrainingMode.UNTIL_LEVEL -> {
+                val dailyExp = tooltip.getDailyExp() ?: return
+                val duration = tooltip.getDuration() ?: return
+                val totalExp = dailyExp * duration
+                val coinPerExp = tooltip.getCoins() / totalExp
+                val xpPerBit = totalExp / tooltip.getBits()
+                tooltip.insertLineAfter(coinsPattern, "§6 ➜ Coins/XP: ${coinPerExp.roundTo(2)}")
+                tooltip.insertLineAfter(bitsPattern, "§b ➜ XP/Bit: ${xpPerBit.roundTo(2)}")
+
+            }
+        }
+    }
+
+    @SubscribeEvent
+    fun onInventoryUpdate(event: InventoryUpdatedEvent) {
+        if (!trainingSlotInventoryPattern.matches(event.inventoryName.removeColor())) return
+        val slot24 = event.inventoryItems[24] ?: return
+
+        val name = slot24.displayName.removeColor()
+        if (desiredLevelPatter.matches(name)) {
+            trainingMode = TrainingMode.UNTIL_LEVEL
+        } else if (userInputPattern.matches(name)) {
+            trainingMode = TrainingMode.DAY_COUNT
+        }
+
+    }
+
+    private fun MutableList<String>.insertLineAfter(pattern: Pattern, content: String) {
+        val iter = this.listIterator()
+        while (iter.hasNext()) {
+            val line = iter.next().removeColor()
+            if (pattern.matcher(line).find()) {
+                iter.add(content)
+            }
+        }
+    }
+
+    // To avoid duplicate code
+    private fun <T> Pattern.read(lore: List<String>, func: (String) -> T): T? {
```

Comment:
Please do not work with the group number of patterns, rather use the group name instead.